django/db/models/query.py

```diff
@@ -806,7 +806,10 @@ def bulk_update(self, objs, fields, batch_size=None):
         if not fields:
             raise ValueError("Field names must be given to bulk_update().")
         objs = tuple(objs)
-        if any(obj.pk is None for obj in objs):
+        pks = set(obj.pk for obj in objs)
+        if len(pks) < len(objs):
+            raise ValueError("bulk_update() cannot update duplicates.")
+        if None in pks:
```

Comment:
What do you think about using `Counter()`?
```python
counter = Counter(obj.pk for obj in objs)
if counter.get(None):
    raise ValueError("All bulk_update() objects must have a primary key set.")
if any(count > 1 for count in counter.values()):
    raise ValueError("...")
```
... or to fail-fast, e.g.
```python
pks = set()
for obj in objs:
    if obj.pk is None:
        raise ValueError("All bulk_update() objects must have a primary key set.")
    if obj.pk in pks:
        raise ValueError(...)
    pks.add(obj.pk) 
```

Reply:
Hmm, imho set serialization and key item assignment (as for dicts) have slightly different mechanics. Not sure, which one works better here. Doesnt key assignment have the risk of falling back to the `hash(value)` evaluation for complex types in cpython, that dont implement `__eq__` properly? Thus it might happily assign another slot on `Counter`, while it reduces to the same value for the db? These equality differences in python vs. SQL might screw up the whole idea (well there is still a comparison closer to SQL possible by using db-prepared values, rather than the python types directly).

Maybe I am exaggerating here and its not an issue at all (because both will work for standard field types*), then I'd lean more toward your fast failing idea.

At least I get the idea, that this tiny sanity check pulls much more into the equation...

---

[*] Still might need some docs, how to build a proper python field type, that also passes the set serialization/key assignment correctly. Otherwise 3rd party field types as pk might lead to weird errors with `bulk_update`.

Reply:
> Hmm, imho set serialization and key item assignment (as for dicts) have slightly different mechanics. Not sure, which one works better here. Doesnt key assignment have the risk of falling back to the hash(value) (identity) evaluation for complex types in cpython, that dont implement __eq__ properly (equality)?

Both `set` and `set` rely on the same `__hash__` logic you are referring to so if you are worried Python/SQL incoherence about using `Counter` you should be equally worried about using `set`.

The problem is the same as ticket-25544 though (see 86eccdc8b67728d84440a46e5bf62c78f2eddf6d) it's possible that `model.pk` is not hashable and we must account for it one way or another.

Reply:
Yes I agree, for correctness `set` isn't any better, just different. Will have to look up, how the CASE chaining itself deals with it. (Does it just repeat values for duplicate pks and let the db decide? Then this might not be "stable" as well...)